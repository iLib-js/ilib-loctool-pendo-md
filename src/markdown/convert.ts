import fromMarkdown from "mdast-util-from-markdown";
import strikethrough from "mdast-util-gfm-strikethrough";
import toMarkdown from "mdast-util-to-markdown";
import strikethroughSyntax from "micromark-extension-gfm-strikethrough";
import unistUtilRemovePosition from "unist-util-remove-position";
import underline from "./micromark-plugin/underline";
import colorString from "./string-transformer/color";
import colorAst from "./ast-transformer/color";
import unsupported from "./micromark-plugin/unsupported";

import type { Root } from "mdast";

export const parse = (markdown: string): Root => {
    // transform Pendo markdown Color tags to HTML tags which can later be processed through AST transformations
    const markdownWithColorAsHtml = colorString.toHtmlTags(markdown);

    // parse the markdown with the strikethrough and underline syntax extensions
    // (color nodes will show up as HTML nodes in the AST)
    const ast = fromMarkdown(markdownWithColorAsHtml, {
        extensions: [unsupported.syntax, strikethroughSyntax({ singleTilde: false }), underline.syntax()],
        mdastExtensions: [strikethrough.fromMarkdown, underline.mdastEstension.fromMarkdown],
    });

    // remove position data from the AST since it's not needed
    const astNoPosition = unistUtilRemovePosition(ast) as Root;

    // transform the AST to replace the HTML nodes with custom Color nodes
    const astWithColorNodes = colorAst.toColorNodes(astNoPosition);

    return astWithColorNodes;
};

export const stringify = (tree: Root): string => {
    // backconvert the custom Color nodes to HTML nodes
    const astNoColorNodes = colorAst.fromColorNodes(tree);

    // convert the AST to markdown with the strikethrough and underline syntax extensions
    // (unsupported syntax will be left as is since it is embedded in the AST as plain text)
    const markdown = toMarkdown(astNoColorNodes, {
        // note: we're passing mdast extensions here even though the name is `extensions` which in `fromMarkdown` is used for syntax extensions
        // this is verbatim from https://github.com/syntax-tree/mdast-util-gfm-strikethrough/tree/0.2.3?tab=readme-ov-file#use
        // and also fine according to the types
        extensions: [strikethrough.toMarkdown, underline.mdastEstension.toMarkdown],
    });

    // backconvert the HTML tags to Pendo markdown Color tags
    const markdownWithColorTags = colorString.fromHtmlTags(markdown);

    return markdownWithColorTags;
};
