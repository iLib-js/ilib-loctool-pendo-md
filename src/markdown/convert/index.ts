export { convert, backconvert } from "./convert";
