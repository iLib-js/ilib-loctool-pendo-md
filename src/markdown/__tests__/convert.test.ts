import { parse } from "../convert";

describe("Pendo Markdown Parser", () => {
    describe("basic markdown", () => {
        // basic markdown syntax which Pendo supports
        // i.e. bold, italic, link, ordered list (numbers), and unordered list (dash, asterisk, plus)
        // as per https://support.pendo.io/hc/en-us/articles/360031866552-Use-markdown-syntax-for-guide-text-styling
        it.each([
            ["bold", "**bold**"],
            ["italic (asterisk)", "*italic (asterisk)*"],
            ["italic (underscore)", "_italic (underscore)_"],
            ["link", "[link](https://example.com)"],
            ["ordered list", "1. ordered list"],
            ["unordered list (dash)", "- unordered list item 1\n- unordered list item 2\n- unordered list item 3"],
            ["unordered list (asterisk)", "* unordered list item 1\n* unordered list item 2\n* unordered list item 3"],
            ["unordered list (plus)", "+ unordered list item 1\n+ unordered list item 2\n+ unordered list item 3"],
        ])(`parses %s`, (_, value) => {
            const ast = parse(value);
            expect(ast).toMatchSnapshot();
        });
    });

    // custom Pendo extensions for markdown syntax
    // i.e. strikethrough, underline, and color
    // as per https://support.pendo.io/hc/en-us/articles/360031866552-Use-markdown-syntax-for-guide-text-styling
    describe("Pendo extended syntax", () => {
        describe("external libs", () => {
            // GFM-like strikethrough: `~~strikethrough~~`
            it("strikethrough", () => {
                const ast = parse("~~strikethrough~~");
                expect(ast).toMatchSnapshot();
            });
        });

        describe("custom extensions", () => {
            // underline: `++underline++`
            it("parses underline", () => {
                const ast = parse("++underline++");
                expect(ast).toMatchSnapshot();
            });

            // color: `{color: #000000}black{/color}`
            it.skip("parses color", () => {
                const ast = parse("{color: #000000}black{/color}");
                expect(ast).toMatchSnapshot();
            });
        });
    });
});
