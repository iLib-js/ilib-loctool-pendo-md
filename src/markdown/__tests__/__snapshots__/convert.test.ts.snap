// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pendo Markdown Parser Pendo extended syntax custom extensions parses underline 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 12,
                  "line": 1,
                  "offset": 11,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "text",
              "value": "underline",
            },
          ],
          "position": {
            "end": {
              "column": 14,
              "line": 1,
              "offset": 13,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "underline",
        },
      ],
      "position": {
        "end": {
          "column": 14,
          "line": 1,
          "offset": 13,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 14,
      "line": 1,
      "offset": 13,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser Pendo extended syntax external libs strikethrough 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 16,
                  "line": 1,
                  "offset": 15,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "text",
              "value": "strikethrough",
            },
          ],
          "position": {
            "end": {
              "column": 18,
              "line": 1,
              "offset": 17,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "delete",
        },
      ],
      "position": {
        "end": {
          "column": 18,
          "line": 1,
          "offset": 17,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 18,
      "line": 1,
      "offset": 17,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser basic markdown parses bold 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 7,
                  "line": 1,
                  "offset": 6,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "text",
              "value": "bold",
            },
          ],
          "position": {
            "end": {
              "column": 9,
              "line": 1,
              "offset": 8,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "strong",
        },
      ],
      "position": {
        "end": {
          "column": 9,
          "line": 1,
          "offset": 8,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 9,
      "line": 1,
      "offset": 8,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser basic markdown parses italic (asterisk) 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 19,
                  "line": 1,
                  "offset": 18,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "italic (asterisk)",
            },
          ],
          "position": {
            "end": {
              "column": 20,
              "line": 1,
              "offset": 19,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "emphasis",
        },
      ],
      "position": {
        "end": {
          "column": 20,
          "line": 1,
          "offset": 19,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 20,
      "line": 1,
      "offset": 19,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser basic markdown parses italic (underscore) 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 21,
                  "line": 1,
                  "offset": 20,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "italic (underscore)",
            },
          ],
          "position": {
            "end": {
              "column": 22,
              "line": 1,
              "offset": 21,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "emphasis",
        },
      ],
      "position": {
        "end": {
          "column": 22,
          "line": 1,
          "offset": 21,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser basic markdown parses link 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 6,
                  "line": 1,
                  "offset": 5,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "link",
            },
          ],
          "position": {
            "end": {
              "column": 28,
              "line": 1,
              "offset": 27,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "title": null,
          "type": "link",
          "url": "https://example.com",
        },
      ],
      "position": {
        "end": {
          "column": 28,
          "line": 1,
          "offset": 27,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 28,
      "line": 1,
      "offset": 27,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser basic markdown parses ordered list 1`] = `
{
  "children": [
    {
      "children": [
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 16,
                      "line": 1,
                      "offset": 15,
                    },
                    "start": {
                      "column": 4,
                      "line": 1,
                      "offset": 3,
                    },
                  },
                  "type": "text",
                  "value": "ordered list",
                },
              ],
              "position": {
                "end": {
                  "column": 16,
                  "line": 1,
                  "offset": 15,
                },
                "start": {
                  "column": 4,
                  "line": 1,
                  "offset": 3,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 16,
              "line": 1,
              "offset": 15,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": true,
      "position": {
        "end": {
          "column": 16,
          "line": 1,
          "offset": 15,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "spread": false,
      "start": 1,
      "type": "list",
    },
  ],
  "position": {
    "end": {
      "column": 16,
      "line": 1,
      "offset": 15,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser basic markdown parses unordered list (asterisk) 1`] = `
{
  "children": [
    {
      "children": [
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 1,
                      "offset": 23,
                    },
                    "start": {
                      "column": 3,
                      "line": 1,
                      "offset": 2,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 1",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 1,
                  "offset": 23,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 1,
              "offset": 23,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 2,
                      "offset": 47,
                    },
                    "start": {
                      "column": 3,
                      "line": 2,
                      "offset": 26,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 2",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 2,
                  "offset": 47,
                },
                "start": {
                  "column": 3,
                  "line": 2,
                  "offset": 26,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 2,
              "offset": 47,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 24,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 3,
                      "offset": 71,
                    },
                    "start": {
                      "column": 3,
                      "line": 3,
                      "offset": 50,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 3",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 3,
                  "offset": 71,
                },
                "start": {
                  "column": 3,
                  "line": 3,
                  "offset": 50,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 3,
              "offset": 71,
            },
            "start": {
              "column": 1,
              "line": 3,
              "offset": 48,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": {
        "end": {
          "column": 24,
          "line": 3,
          "offset": 71,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": {
    "end": {
      "column": 24,
      "line": 3,
      "offset": 71,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser basic markdown parses unordered list (dash) 1`] = `
{
  "children": [
    {
      "children": [
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 1,
                      "offset": 23,
                    },
                    "start": {
                      "column": 3,
                      "line": 1,
                      "offset": 2,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 1",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 1,
                  "offset": 23,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 1,
              "offset": 23,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 2,
                      "offset": 47,
                    },
                    "start": {
                      "column": 3,
                      "line": 2,
                      "offset": 26,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 2",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 2,
                  "offset": 47,
                },
                "start": {
                  "column": 3,
                  "line": 2,
                  "offset": 26,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 2,
              "offset": 47,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 24,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 3,
                      "offset": 71,
                    },
                    "start": {
                      "column": 3,
                      "line": 3,
                      "offset": 50,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 3",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 3,
                  "offset": 71,
                },
                "start": {
                  "column": 3,
                  "line": 3,
                  "offset": 50,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 3,
              "offset": 71,
            },
            "start": {
              "column": 1,
              "line": 3,
              "offset": 48,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": {
        "end": {
          "column": 24,
          "line": 3,
          "offset": 71,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": {
    "end": {
      "column": 24,
      "line": 3,
      "offset": 71,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`Pendo Markdown Parser basic markdown parses unordered list (plus) 1`] = `
{
  "children": [
    {
      "children": [
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 1,
                      "offset": 23,
                    },
                    "start": {
                      "column": 3,
                      "line": 1,
                      "offset": 2,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 1",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 1,
                  "offset": 23,
                },
                "start": {
                  "column": 3,
                  "line": 1,
                  "offset": 2,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 1,
              "offset": 23,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 2,
                      "offset": 47,
                    },
                    "start": {
                      "column": 3,
                      "line": 2,
                      "offset": 26,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 2",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 2,
                  "offset": 47,
                },
                "start": {
                  "column": 3,
                  "line": 2,
                  "offset": 26,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 2,
              "offset": 47,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 24,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 24,
                      "line": 3,
                      "offset": 71,
                    },
                    "start": {
                      "column": 3,
                      "line": 3,
                      "offset": 50,
                    },
                  },
                  "type": "text",
                  "value": "unordered list item 3",
                },
              ],
              "position": {
                "end": {
                  "column": 24,
                  "line": 3,
                  "offset": 71,
                },
                "start": {
                  "column": 3,
                  "line": 3,
                  "offset": 50,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 24,
              "line": 3,
              "offset": 71,
            },
            "start": {
              "column": 1,
              "line": 3,
              "offset": 48,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": {
        "end": {
          "column": 24,
          "line": 3,
          "offset": 71,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": {
    "end": {
      "column": 24,
      "line": 3,
      "offset": 71,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
